File: sorts/Shell.java
package sorts;

public class Shell {

	public static void sort(int[] arr) {
		int n = arr.length;
		int i = 1;
		while (i < n / 3) {
			i = 3 * i + 1;
		}
		while (i >= 1) {
			for (int j = i; j < n; j++) {
				for (int k = j; k >= i && arr[k] < arr[k - i]; k -= i) {
					int o = arr[k];
					arr[k] = arr[k - i];
					arr[k - i] = o;
				}
			}
			i = i / 3;
		}
	}

}

File: sorts/shell.go
package sorts

//ShellSort []int
func ShellSort(arr []int) {
	n := len(arr)
	i := 1
	for i < n/3 {
		i = 3*i + 1
	}
	for i >= 1 {
		for j := i; j < n; j++ {
			for k := j; k >= i && arr[k] < arr[k-i]; k -= i {
				arr[k], arr[k-i] = arr[k-i], arr[k]
			}
		}
		i = i / 3
	}
}
